packages/ckeditor5-engine/src/view/observer/inputobserver.ts

```diff
@@ -68,32 +70,50 @@ export default class InputObserver extends DomEventObserver<'beforeinput'> {
 		// in the fake selection container.
 		if ( viewDocument.selection.isFake ) {
 			// Future-proof: in case of multi-range fake selections being possible.
-			targetRanges = [ ...viewDocument.selection.getRanges() ];
+			targetRanges = Array.from( viewDocument.selection.getRanges() );
 
 			// @if CK_DEBUG_TYPING // if ( window.logCKETyping ) {
 			// @if CK_DEBUG_TYPING // 	console.info( '%c[InputObserver]%c using fake selection:',
-			// @if CK_DEBUG_TYPING // 		'color: green;font-weight: bold', 'font-weight:bold', targetRanges
+			// @if CK_DEBUG_TYPING // 		'color: green;font-weight: bold', 'font-weight:bold', targetRanges,
+			// @if CK_DEBUG_TYPING // 		viewDocument.selection.isFake ? 'fake view selection' : 'fake DOM parent'
 			// @if CK_DEBUG_TYPING // 	);
 			// @if CK_DEBUG_TYPING // }
-		} else {
+		} else if ( domTargetRanges.length ) {
 			targetRanges = domTargetRanges.map( domRange => {
-				return view.domConverter.domRangeToView( domRange );
+				return view.domConverter.domRangeToView( domRange )!;
 			} );
 
 			// @if CK_DEBUG_TYPING // if ( window.logCKETyping ) {
 			// @if CK_DEBUG_TYPING // 	console.info( '%c[InputObserver]%c using target ranges:',
 			// @if CK_DEBUG_TYPING // 		'color: green;font-weight: bold', 'font-weight:bold', targetRanges
 			// @if CK_DEBUG_TYPING // 	);
 			// @if CK_DEBUG_TYPING // }
+		} else if ( env.isAndroid ) {
+			const domSelection = ( domEvent.target as HTMLElement ).ownerDocument.defaultView!.getSelection()!;
+
+			targetRanges = Array.from( view.domConverter.domSelectionToView( domSelection ).getRanges() );
+
+			// @if CK_DEBUG_TYPING // if ( window.logCKETyping ) {
+			// @if CK_DEBUG_TYPING // 	console.info( '%c[InputObserver]%c using selection ranges:',
+			// @if CK_DEBUG_TYPING // 		'color: green;font-weight: bold', 'font-weight:bold', targetRanges
+			// @if CK_DEBUG_TYPING // 	);
+			// @if CK_DEBUG_TYPING // }
 		}
 
-		this.fire( domEvent.type, domEvent, {
-			data,
-			dataTransfer,
-			isComposing: domEvent.isComposing,
-			targetRanges,
-			inputType: domEvent.inputType
-		} );
+		if ( env.isAndroid && domEvent.inputType == 'insertCompositionText' && data && data.endsWith( '\n' ) ) {
+			this.fire( domEvent.type, domEvent, {
+				inputType: 'insertParagraph',
+				targetRanges: [ view.createRange( targetRanges[ 0 ].end ) ]
+			} );
+		} else {
```

Comment:
I don't like that the default (expected) behavior of this listener is in an `else{}` block because it made it harder for me to understand that the `if()` part is an exception and then there goes the default.

More naturally for me it would be like this:

```js
if ( specialCondition ) {
   doSpecialThings();
   return;
}

doDefaultThings();
```

However, then, there's the `console.groupEnd()` thing, so we can't actually make it an early return (because we'd miss the console thing).

So... maybe just document the `else{}` block:

```js
// Special case for Android: ...
if( ) {
   ...
}
// Default action: Fire (what do we actually fire here?! Document what it could bew)
else {
   ...
}
```